File: Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/ClientApi/Operations/Server/ClientCertificate.cs
using System.Collections.Generic;
using Raven.Client.Documents;
using Raven.Client.ServerWide.Operations;
using Raven.Client.ServerWide.Operations.Certificates;

namespace Raven.Documentation.Samples.ClientApi.Operations.Server
{

    public class ClientCertificate
    {
        private interface IFoo
        {
            /*
            #region cert_1
            public CreateClientCertificateOperation(string name, Dictionary<string, DatabaseAccess> permissions, SecurityClearance clearance, string password = null)
            #endregion
            */
        }

        private class Foo
        {
            #region cert_2
            public enum SecurityClearance
            {
                ClusterAdmin,
                ClusterNode,
                Operator,
                ValidUser
            }
            #endregion

            #region cert_3
            public enum DatabaseAccess
            {
                ReadWrite,
                Admin
            }
            #endregion
        }

        public ClientCertificate()
        {
            using (var store = new DocumentStore())
            {
                {
                    #region cert_4
                    // With user role set to Cluster Administator or Operator the user of this certificate 
                    // is going to have access to all databases
                    CreateClientCertificateOperation operation = new CreateClientCertificateOperation("admin", null, SecurityClearance.Operator);
                    CertificateRawData certificateRawData = store.Maintenance.Server.Send(operation);
                    byte[] cert = certificateRawData.RawData;
                    #endregion
                }

                {
                    #region cert_5
                    // when security clearance is ValidUser, you need to specify per database permissions
                    CreateClientCertificateOperation operation = new CreateClientCertificateOperation("user1", new Dictionary<string, DatabaseAccess>
                    {
                        { "Northwind", DatabaseAccess.Admin }
                    }, SecurityClearance.ValidUser, "myPassword");
                    CertificateRawData certificateRawData = store.Maintenance.Server.Send(operation);
                    byte[] cert = certificateRawData.RawData;
                    #endregion
                }
            }
        }
    }
}

File: Documentation/5.2/Samples/java/src/test/java/net/ravendb/ClientApi/Operations/Server/ClientCertificate.java
package net.ravendb.ClientApi.Operations.Server;

import net.ravendb.client.documents.DocumentStore;
import net.ravendb.client.documents.IDocumentStore;
import net.ravendb.client.serverwide.operations.certificates.*;
import org.apache.commons.codec.binary.Base64;

import java.security.KeyStore;
import java.security.cert.Certificate;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

public class ClientCertificate {

    private interface IFoo {
        /*
        //region cert_1_1
        public CreateClientCertificateOperation(String name,
                                                Map<String, DatabaseAccess> permissions,
                                                SecurityClearance clearance)

        public CreateClientCertificateOperation(String name,
                                                Map<String, DatabaseAccess> permissions,
                                                SecurityClearance clearance,
                                                String password)
        //endregion
        */

        /*
        //region get_cert_1
        public GetCertificateOperation(String thumbprint)
        //endregion
        */

        /*
        //region get_certs_1
        public GetCertificatesOperation(int start, int pageSize)
        //endregion
        */

        /*
        //region cert_put_1
        public PutClientCertificateOperation(String name,
                                             String certificate,
                                             Map<String, DatabaseAccess> permissions,
                                             SecurityClearance clearance)
        //endregion
        */
    }

    private static class Foo {
        //region cert_1_2
        public enum SecurityClearance {
            CLUSTER_ADMIN,
            CLUSTER_NODE,
            OPERATOR,
            VALID_USER
        }
        //endregion

        //region cert_1_3
        public enum DatabaseAccess {
            READ,
            READ_WRITE,
            ADMIN
        }
        //endregion
    }


    public ClientCertificate() {
        try (IDocumentStore store = new DocumentStore()) {
            //region cert_1_4
            // With user role set to Cluster Administrator or Operator the user of this certificate
            // is going to have access to all databases

            CreateClientCertificateOperation operation = new CreateClientCertificateOperation("admin",
                null, SecurityClearance.OPERATOR);
            CertificateRawData certificateRawData = store.maintenance().server().send(operation);
            byte[] certificatesZipped = certificateRawData.getRawData();
            //endregion
        }

        try (IDocumentStore store = new DocumentStore()) {
            //region cert_1_5
            // when security clearance is ValidUser, you need to specify per database permissions
            CreateClientCertificateOperation operation = new CreateClientCertificateOperation("user1",
                Collections.singletonMap("Northwind", DatabaseAccess.ADMIN),
                SecurityClearance.VALID_USER,
                "myPassword");

            CertificateRawData certificateRawData = store.maintenance().server().send(operation);
            byte[] certificateZipped = certificateRawData.getRawData();
            //endregion
        }

        try (IDocumentStore store = new DocumentStore()) {
            //region get_cert_2
            String thumbprint = "a909502dd82ae41433e6f83886b00d4277a32a7b";
            CertificateDefinition definition = store.maintenance()
                .server()
                .send(new GetCertificateOperation(thumbprint));
            //endregion
        }

        try (IDocumentStore store = new DocumentStore()) {
            //region get_certs_2
            CertificateDefinition[] definitions = store.maintenance()
                .server()
                .send(new GetCertificatesOperation(0, 20));
            //endregion
        }

        try (IDocumentStore store = new DocumentStore()) {
            KeyStore keyStore = null;

            try {
                //region cert_put_2
                byte[] rawCert = Files.readAllBytes(Paths.get("<path-to-certificate.crt>"));
                String certificateAsBase64 = Base64.getEncoder().encodeToString(rawCert);                

                store.maintenance().server().send(
                    new PutClientCertificateOperation(
                        "certificateName",
                        certificateAsBase64,
                        new HashMap<>(),
                        SecurityClearance.CLUSTER_ADMIN));
                //endregion
            } catch (Exception e) {
            }
        }
    }
}
